### ISSUE TITLE
Native-Image: Newly created project with Tika extension can't extract anything 

### ISSUE THREAD

**Issue by pelletier197** [quarkusio/quarkus#10104]
**Describe the bug**
I tried creating a Tika project following [the documentation](https://quarkus.io/guides/tika) on Quarkus. I think any pdf would produce the same result, but here's the one I tried.
[americanexpress_01.pdf](https://github.com/quarkusio/quarkus/files/4801495/americanexpress_01.pdf)

**Expected behavior**
To give a little context, I should be able to extract any text from any document tika supports, so I simply need to have the whole Tika library bundle in the native application, which doesn't seem to be what is happening right now.

**Actual behavior**
With the documentation's example, when doing the curl post with the PDF above, I get.
```
Error: Could not find referenced cmap stream Identity-H
```

Looking at the stacktrace, I realise that this resource is missing inside PDFBox (a tika dependency) inside the bundled native-image, along with all other tika dependencies... 

**Configuration**
```properties
quarkus.package.type=native
quarkus.package.uber-jar=true
quarkus.log.console.enable=false
quarkus.native.add-all-charsets=true
quarkus.native.additional-build-args=-H:ReflectionConfigurationFiles=reflection-config.json,-H:ResourceConfigurationFiles=resources-config.json 
```

**What I tried**
To test really fast I tried adding a `resources-config.json` to the project.
```
{
  "resources": [
    {
      "pattern": ".*"
    }
  ]
}
```
and now success, the native app can extract the text of my file (yeah). But, trying different PDFs leaded to different errors, and this time they were runtime errors.

So, considering I followed the documentation properly and that the native-image build process ends without any error, I'm really wondering what i'm doing wrong, since this is supposed to parse any files and it doesn't, because there seems to be missing A LOT of dependencies inside the native-image.

Is the Quarkus Tika extension really supposed to be usable, or do I have to create the extension with all I need myself as done in the [TikaProcessor.java](https://github.com/quarkusio/quarkus/blob/master/extensions/tika/deployment/src/main/java/io/quarkus/tika/deployment/TikaProcessor.java)  (which doesn't seem to include many things)

For my use-case, is there a way I can include everything from Tika ?

**Comment by sberyozkin** [comment]
Hi @pelletier197 Thanks for opening this issue. PDF files are being tested in the native mode, but as your example shows, there could be some code path inside the Tika PDF parser or likely PDFBox which need more native image support. 
I'm assigning this issue to myself but I don't help when I can get it fixed, if you can find some time to investigate then it would be appreciated. 

Do you know what `Could not find referenced cmap stream Identity-H` can mean ? is there some resource file/service provder file somewhere in PDFBox for example which may need to be included ?
It should be a straightforward fix, we just need to figure out where this `Identity-H` thing is coming from.
Can you debug in a JVM mode and see how this `Identity-H` is being loaded ?

Cheers

**Comment by sberyozkin** [comment]
@pelletier197 These are FontBox resources:
https://svn.apache.org/viewvc/pdfbox/trunk/fontbox/src/main/resources/org/apache/fontbox/cmap/

So all we need is 
```
@BuildStep
public void registerFontBoxResources(BuildProducer<NativeImageResourceBuildItem> resource) {
        resource.produce(new NativeImageResourceBuildItem("org/apache/fontbox/cmap/*"));
        resource.produce(new NativeImageResourceBuildItem("org/apache/fontbox/unicode/Scripts.txt"));
}
```

I need to clarify how to load all of them fast under a singe dir...

**Comment by pelletier197** [comment]
@sberyozkin yes exactly I had figured that out doing my research. 

However, I don't know if regexs would work with paths since I haven't found any example with a full path in the Graalvm documentation. 

I can try to figure out what exactly is the regex needed to match those resources, but it will not be enough to solve my problem, since tika instantiate a lot of things by reflection that are not being loaded. 

For instance *xslx* files did not seem to have a parser. 

Some other type of pdf raise other kind of errors. There is just so many cases I can't test all of them to see what's not loaded, considering the build takes 5 minutes every time.

**Comment by sberyozkin** [comment]
@pelletier197, if you can attach the test PDF files with the corresponding errors then we can tackle them one by one. This is really the only way forward with or without Quarkus, for Tika to work in the native image, and its PDF parser in particular. 
> For instance xslx files did not seem to have a parser.
The parser must be included, a few of them are excluded from the native imafe but certainly not `xslx` ones. It could be for example an media-type detection issue for that specific file in the native image, so if you can, please attach that file too

**Comment by pelletier197** [comment]
Sadly, I cannot provide you with the PDFs for security reasons. I can however do better than that. 

Since I really need this thing to work with GraalVM, I will use the [tracing agent](https://github.com/oracle/graal/blob/master/substratevm/CONFIGURE.md) provided by graalVM to generate the configuration files based on a 40Gb dataset I have (not enough for all the possible files, but better than nothing). 

I'll be happy to provide you with the generated files for reflection and resources configuration if that helps you :smiley:

**Comment by sberyozkin** [comment]
Hi @pelletier197 thanks, good you have a solution for now. 
OK, let me first try to fix this particular issue, I think I will get the time before 1.6.0. 
I'll ping you here once it is done to see what can be done - without the test files it can be hard to get any assurance adding a given resource will work. Seeing a generated resource set can help I guess :-) It should be not large I'd say...
Cheers

**Comment by pelletier197** [comment]
If that can help you, here are the generated configurations for this only file I sent you.

[dependencies.zip](https://github.com/quarkusio/quarkus/files/4805339/dependencies.zip)

Thank you for your feedback!

**Comment by sberyozkin** [comment]
@pelletier197 thanks for your help :-)

**Comment by borellda** [comment]
Hello,

I need to trigger this one.
I'm creating a small service to download files and extract text from it. And of course I would like to use a native image.
However, using Quarkus 1.6.1 and GralalVM 20.1.0 (Java 11.0.7), for this file:
[data_viewer_manual_nec_v10.pdf](https://github.com/quarkusio/quarkus/files/5118333/data_viewer_manual_nec_v10.pdf)
I receive the following :
`15:16:52 INFO  [or.ar.se.TextExtractor] (vert.x-eventloop-thread-1) Extract text for file
15:16:52 WARN  [or.ap.pd.pd.fo.PDTrueTypeFont] (vert.x-eventloop-thread-1) Using fallback font 'LiberationSans' for 'Arial-BoldMT'
15:16:52 WARN  [or.ap.pd.pd.fo.PDTrueTypeFont] (vert.x-eventloop-thread-1) Using fallback font 'LiberationSans' for 'ArialMT'
15:16:52 WARN  [or.ap.pd.pd.fo.PDTrueTypeFont] (vert.x-eventloop-thread-1) Using fallback font 'LiberationSans' for 'TimesNewRomanPSMT'
15:16:52 WARN  [or.ap.pd.pd.fo.PDTrueTypeFont] (vert.x-eventloop-thread-1) Using fallback font 'LiberationSans' for 'Arial-ItalicMT'
15:16:52 WARN  [or.ap.pd.pd.fo.PDTrueTypeFont] (vert.x-eventloop-thread-1) Using fallback font 'LiberationSans' for 'Arial-BoldItalicMT'
15:16:53 WARN  [or.ar.se.TextExtractor] (vert.x-eventloop-thread-1) io.quarkus.tika.TikaParseException: Failed to extract text`

In JVM mode it works.
Any chance there is a fix already in preparation?

Regards,
Daniel

**Comment by pelletier197** [comment]
> Hello,
> 
> I need to trigger this one.
> I'm creating a small service to download files and extract text from it. And of course I would like to use a native image.
> However, using Quarkus 1.6.1 and GralalVM 20.1.0 (Java 11.0.7), for this file:
> [data_viewer_manual_nec_v10.pdf](https://github.com/quarkusio/quarkus/files/5118333/data_viewer_manual_nec_v10.pdf)
> I receive the following :
> `15:16:52 INFO [or.ar.se.TextExtractor] (vert.x-eventloop-thread-1) Extract text for file 15:16:52 WARN [or.ap.pd.pd.fo.PDTrueTypeFont] (vert.x-eventloop-thread-1) Using fallback font 'LiberationSans' for 'Arial-BoldMT' 15:16:52 WARN [or.ap.pd.pd.fo.PDTrueTypeFont] (vert.x-eventloop-thread-1) Using fallback font 'LiberationSans' for 'ArialMT' 15:16:52 WARN [or.ap.pd.pd.fo.PDTrueTypeFont] (vert.x-eventloop-thread-1) Using fallback font 'LiberationSans' for 'TimesNewRomanPSMT' 15:16:52 WARN [or.ap.pd.pd.fo.PDTrueTypeFont] (vert.x-eventloop-thread-1) Using fallback font 'LiberationSans' for 'Arial-ItalicMT' 15:16:52 WARN [or.ap.pd.pd.fo.PDTrueTypeFont] (vert.x-eventloop-thread-1) Using fallback font 'LiberationSans' for 'Arial-BoldItalicMT' 15:16:53 WARN [or.ar.se.TextExtractor] (vert.x-eventloop-thread-1) io.quarkus.tika.TikaParseException: Failed to extract text`
> 
> In JVM mode it works.
> Any chance there is a fix already in preparation?
> 
> Regards,
> Daniel

I don't know on their side how this is going, but as I said, I found a workaround by running the project with the graalvm tracing agent that detects what classes and resources are instantiated and used at runtime and generates a configuration file you can use to build your native image. 

It makes a great overall job, but there will still be missing a few dependencies you'll need to add manually. 

Tika is really not written to be compatible with GrallVM for now, but that will hopefully change eventually.

**Comment by borellda** [comment]
Thank you for the quick reply.

Could you point me to an entry point on how to do that? The link you provided above seems to be dead.

Regards,
Daniel

**Comment by pelletier197** [comment]
You're right, the link seems dead. There is [this medium post](https://medium.com/graalvm/introducing-the-tracing-agent-simplifying-graalvm-native-image-configuration-c3b56c486271#:~:text=tl%3Bdr%3A%20The%20tracing%20agent,%2C%20resource%2C%20and%20proxy%20usage) that explains well the base configuration.

**Comment by sberyozkin** [comment]
Hi @pelletier197 and @borellda 

I'm sorry it has taken me so long to fix this issue, the PR has just been opened, #12484. It is hard for me to prioritize on Tika issues at the moment.
@pelletier197 #12484 will close this issue - but I'd appreciate if you could check the fix later on against your file set, you mentioned there was another file or format which was not processed correctly - if you could open a new issue with a simplified failing file then it would be good.
Thanks and again sorry for a delay

**Comment by sberyozkin** [comment]
@pelletier197, hi, I'm assuming the PDF file linked here is a test file which does not point to any concrete data ? Can you confirm it please ? I can ask in the Tika forum for a small test PDF file instead if it is problematic. CC @gastaldi

**Comment by sberyozkin** [comment]
@pelletier197 I've asked Tika guys for the alternative file just in case...

**Comment by pelletier197** [comment]
The PDF I have here is just a simple PDF with the minimal information to reproduce. 

I you want, Tika has a test folder with a lot of files they support. I would be really great if quarkus could support all those files out of the box, since I believe this test set covers a lot of their API.

Your can find this test folder at this link https://github.com/apache/tika/tree/master/tika-parsers/src/test/resources/test-documents

**Comment by pelletier197** [comment]
I also have the configuration file I used to support all the test cases I need to support. I could provide you with it, but it's quite enormous. There are hundreds of resources and runtime loaded classes in this config file.

**Comment by sberyozkin** [comment]
@pelletier197 OK, thanks, I'll then go ahead and merge the PR. Yes, I'm aware of that folder, I copied one of the files (renamed it) to `integration-tests/tika` awhile back. I agree a more exhaustive testing would help, but testing a large set as part of the regular Quarkus build may be problematic. For now I'm trying to address the issues as they are reported though I'm late on a few of them (POI is not happy in the native mode in particuar).
Thanks
